### PR TITLE
Use custom coalescing heuristic and introduce a number of new configuration options for memory pools

### PR DESCRIPTION
**Pull request by msimberg**
This is intended to replace https://github.com/eth-cscs/DLA-Future/pull/1143 as a more permanent solution.

This introduces:
- configuration options for controlling the "next" allocation size (i.e. all blocks after the first one) and alignment size of pools
- a custom coalescing heuristic that works around https://github.com/LLNL/Umpire/issues/906
- configuration options that go along with the custom heuristic

I'm 100% sure I haven't chosen the best names for the new options, but I'd like to discuss them with you first before making any further changes. Additionally, I'm 100% aware that the coalescing options add complexity. My intention with them is primarily to have as debugging options, not necessarily something most users would change. The default behaviour is very "vanilla", but in case we see situations where too much memory is used or similar we can use these options to try to tune how umpire frees/coalesces blocks in the pool.

As a semi-related change I've also namespaced the Umpire pools with a `DLAF_` prefix since we used quite generic names. I suppose we'd notice if the names clash, but I'd rather be on the safe side and avoid clashes (e.g. if the non-thread safe pool is accessed without synchronization I expect bad things to happen).

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
Context for current names: https://github.com/LLNL/Umpire/blob/bf2df8d46f6e8918b7402a45b0f95127c3121478/src/umpire/strategy/QuickPool.hpp#L46-L62.

File: src/init.cpp [hunk 1]
```diff
@@ -129,6 +143,13 @@ struct parseFromString<SizeType> {
   }
 };
 
+template <>
+struct parseFromString<double> {
+  static std::optional<SizeType> call(const std::string& var) {
```
Comment (RMeli) [inline src/init.cpp:148]:
```suggestion
  static std::optional<double> call(const std::string& var) {
```

File: include/dlaf/init.h [hunk 2]
```diff
@@ -38,7 +38,15 @@ struct configuration {
   std::size_t num_np_gpu_streams_per_thread = 3;
   std::size_t num_hp_gpu_streams_per_thread = 3;
   std::size_t umpire_host_memory_pool_initial_bytes = 1 << 30;
+  std::size_t umpire_host_memory_pool_next_bytes = 1 << 30;
```
Comment (msimberg) [inline include/dlaf/init.h:41]: If I were to rename these I'd probably do:
```suggestion
  std::size_t umpire_host_memory_pool_initial_block_bytes = 1 << 30;
  std::size_t umpire_host_memory_pool_next_block_bytes = 1 << 30;
```
Is that an improvement? Other ideas?

**Comment by msimberg** [comment]
cscs-ci run

File: include/dlaf/init.h (hunk 2 above)
Comment (msimberg) [reply include/dlaf/init.h]: I've updated to the suggestion above.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Review by rasolca — APPROVED** [review]

**Pull request merged**